### PR TITLE
Allow url parameter for remote file validation

### PR DESCRIPTION
**Pull request by sebbalex**
this will implements #3

### CODE REVIEW

File: src/main_test.go [hunk 1]
```diff
@@ -119,6 +119,21 @@ func TestValidationWithNetwork(t *testing.T) {
...
 
+func TestValidationRemoteURL(t *testing.T) {
+	out, err := ioutil.ReadFile("../tests/out_valid.minimal.yml")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	urlString := "https://raw.githubusercontent.com/italia/pc-web-validator/master/tests/valid.minimal.yml"
+	// invalid
+	// urlString := "https://raw.githubusercontent.com/italia/pc-web-validator/master/tests/invalid.yml"
```
Comment (libremente) [inline src/main_test.go]: Why not adding also the invalid url test?

File: src/main.go [hunk 2]
```diff
@@ -128,11 +133,60 @@ func (app *App) parse(b []byte) ([]byte, error, error) {
...
+	urlString := vars["url"]
+	if urlString == "" {
+		return
+	}
+
+	//parsing
+	pc, errParse, errConverting := app.parseRemoteURL(urlString)
+
+	if errConverting != nil {
+		log.Errorf("Error converting: %v", errConverting)
```
Comment (libremente) [inline src/main.go]: Are you just logging the error? Nothing is returned?

Comment (sebbalex) [reply src/main.go]: 👍

File: src/main.go [hunk 3]
```diff
@@ -119,7 +129,7 @@ func (app *App) parse(b []byte) ([]byte, error, error) {
 		p.DisableNetwork = app.DisableNetwork
```
[the remark this answers is not in the document]
Comment (libremente) [reply src/main.go]: Also, I don't personally like L119. Check https://stackoverflow.com/questions/52255683/how-to-return-nil-string-in-go maybe it's better to have double return types or use a pointer?

File: src/main.go [hunk 4]
```diff
@@ -128,11 +138,106 @@ func (app *App) parse(b []byte) ([]byte, error, error) {
...
+	if urlString == "" {
+		promptError(errors.New("Not found"), w, http.StatusNotFound, "URL error")
+		return
+	}
+
+	//parsing
+	pc, errParse, errConverting := app.parseRemoteURL(urlString)
+
+	if errConverting != nil {
+		log.Debugf("Error converting: %s", errConverting)
```
Comment (libremente) [inline src/main.go]: I'd probably remove this line since you are already logging inside the `promptError` function, right?

File: src/main.go [hunk 5]
```diff
@@ -128,11 +138,106 @@ func (app *App) parse(b []byte) ([]byte, error, error) {
...
+		promptError(errConverting, w, http.StatusBadRequest, "Error converting")
+	}
+	if errParse != nil {
+		if match, _ := regexp.MatchString(`404`, errParse.Error()); match {
+			log.Debugf("URL Error: %s", errParse)
```
Comment (libremente) [inline src/main.go]: Same as above.

File: src/main.go [hunk 6]
```diff
@@ -128,11 +138,106 @@ func (app *App) parse(b []byte) ([]byte, error, error) {
...
+		Status:  httpStatus,
+		Message: mess,
+		Error:   err.Error(),
+	}
+	log.Debugf("message: %v", message)
+	w.Header().Set("Content-type", "application/json")
+	w.WriteHeader(message.Status)
+	json.NewEncoder(w).Encode(message)
+}
+
```
Comment (libremente) [inline src/main.go]: This two funcs look very similar, I'd try to figure out a way to DRY, what do you think?

File: src/main.go [hunk 7]
```diff
@@ -128,11 +138,106 @@ func (app *App) parse(b []byte) ([]byte, error, error) {
...
+		log.Debugf("Error converting: %s", errConverting)
+		promptError(errConverting, w, http.StatusBadRequest, "Error converting")
+	}
+	if errParse != nil {
+		if match, _ := regexp.MatchString(`404`, errParse.Error()); match {
+			log.Debugf("URL Error: %s", errParse)
+			promptError(errors.New("Not found"), w, http.StatusNotFound, "URL error")
+			return
+		}
+		log.Debugf("Validation Errors: %s", errParse)
```
Comment (libremente) [inline src/main.go]: Same as above. I'd try to print a debug message inside the error wrapper function, what do you think?

File: src/main.go (hunk 4 above)
Comment (libremente) [reply src/main.go]: Also, I think that if errConverting is != nil, pc should be nil so shouldn't it be better to return if this is the case? Is it possible to reach L232 where pc == nil?

File: src/main.go [hunk 8]
```diff
@@ -50,18 +55,18 @@ func (app *App) init() {
 }
 
 func (app *App) initLogger() {
-	f, err := os.OpenFile("app.log", os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
+	_, err := os.OpenFile("app.log", os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
 	if err != nil {
 		log.Fatal(err)
 	}
 	// defer f.Close()
```
[the remark this answers is not in the document]
Comment (sebbalex) [reply src/main.go]: App will use twelve factor logging approach.

File: src/main.go (hunk 4 above)
Comment (sebbalex) [reply src/main.go]: first question: yes, they will be removed. About second one `promptError ` function will return an object that will contain http code `httpStatus`, custom message `mess` and error `err` if occurs.

File: src/main.go (hunk 6 above)
Comment (sebbalex) [reply src/main.go]: yes, I completely agree, I will do some hard refactor in a separate PR.

File: src/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sebbalex) [reply src/main.go]: same as below, actually we have a very contorted flow with different request that require different params and managed by the same code. I will do a heavy refactor in a separate PR.

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# Golang CircleCI 2.0 configuration file
+version: 2
+jobs:
+  build:
+    docker:
+      - image: circleci/golang:1.13
+    steps:
+      - checkout
+      - run: cd src; go test -race; cd ..
```
[the remark this answers is not in the document]
Comment (sebbalex) [reply .circleci/config.yml]: typo :)

**Pull request merged**